### PR TITLE
dladdr() - Retrieve symbol name and address from export table

### PR DESCRIPTION
**Pull request by pali**
Fixes: #87

@rhabacker: please review this change

### CODE REVIEW

**Comment by pali** [comment]
Ok, all tests are passing now.

File: src/dlfcn.c [hunk 1]
```diff
@@ -608,11 +642,16 @@ static void *get_address_from_import_address_table( void *iat, DWORD iat_size, v
 /* Holds module filename */
 static char module_filename[2*MAX_PATH];
 
-static BOOL fill_module_info( void *addr, Dl_info *info )
+static BOOL fill_module_info( HMODULE hModuleImport, void *addr, Dl_info *info )
```
Comment (rhabacker) [inline src/dlfcn.c:645]: As this function does more than filling module info, I would name it fill_info() or similiar.

**Review by rhabacker — CHANGES_REQUESTED** [review]
Looks good so far except the function naming issue.

The current implementation only finds symbols exported from the current executable (in static build mode) or library `dl` (in shared build mode). An extension (in a separate mr or not) would be to search the export tables of all loaded shared libraries.

**Comment by pali** [comment]
> The current implementation only finds symbols exported from the current executable (in static build mode) or library dl (in shared build mode). An extension (in a separate mr or not) would be to search the export tables of all loaded shared libraries.

No, code in this pull request already find symbols in supplied executable / DLL library.

Via GetModuleHandleExA/GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS it retrieves HMODULE of executable/DLL and then it search export tables of this executable/DLL.

In case supplied address it import thunk then it also tries to search import tables of HMODULE which belongs to supplied address.

So I think implementation is now complete.

File: src/dlfcn.c (hunk 1 above)
Comment (pali) [reply src/dlfcn.c]: done!

**Comment by rhabacker** [comment]
> Via GetModuleHandleExA/GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS it retrieves HMODULE of executable/DLL and then it search export tables of this executable/DLL.

I see, thanks for this pointer.

**Comment by rhabacker** [comment]
> Via GetModuleHandleExA/GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS it retrieves HMODULE of executable/DLL and then it search export tables of this executable/DLL.

Then you should add a test for checking an address, which is not in the import table for sure.

**Comment by rhabacker** [comment]
> > Via GetModuleHandleExA/GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS it retrieves HMODULE of executable/DLL and then it search export tables of this executable/DLL.
> 
> Then you should add a test for checking an address, which is not in the import table for sure.

In fact, it looks, that now all checks are fetching  the symbol from the export table :-)

**Comment by pali** [comment]
> In fact, it looks, that now all checks are fetching the symbol from the export table

Hm... it makes sense. If you want to import some symbol then it has to be exported. Otherwise you cannot create/link final executable (or DLL) binary. So probably it it is not required ot check import table at all?

**Comment by rhabacker** [comment]
> > In fact, it looks, that now all checks are fetching the symbol from the export table
> 
> Hm... it makes sense. If you want to import some symbol then it has to be exported. Otherwise you cannot create/link final executable (or DLL) binary. So probably it it is not required ot check import table at all?

It looks so - the GetModuleHandleExA trick is very nice :-)

**Comment by pali** [comment]
I'm not sure if we need to look for symbols in import table in this case...

**Comment by rhabacker** [comment]
> I'm not sure if we need to look for symbols in import table in this case...

I think you can remove that.

**Comment by pali** [comment]
Ok, I have done it, pull request is updated.

**Review by rhabacker — APPROVED** [review]
looks good

**Comment by pali** [comment]
Ok! I think it is ready for merge.

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
Thanks @pali for the PR and @rhabacker for the review!

**Pull request merged**